pkg/scheduler/framework/plugins/dynamicresources/structured/namedresources/namedresourcesmodel_test.go

```diff
@@ -0,0 +1,294 @@
+				Name: instance1,
+			},
+		}},
+	}
+
+	instance2 := "instance-2"
+	twoInstances := Model{
+		Instances: []InstanceAllocation{
+			{
+				Instance: &resourceapi.NamedResourcesInstance{
+					Name: instance1,
+				},
+			},
+			{
+				Instance: &resourceapi.NamedResourcesInstance{
+					Name: instance2,
+				},
+			},
+		},
+	}
+
+	testcases := map[string]struct {
+		model    Model
+		filter   *resourceapi.NamedResourcesFilter
+		requests []*resourceapi.NamedResourcesRequest
+
+		expectCreateErr   string
+		expectAllocation  []string
+		expectAllocateErr string
+	}{
+		"empty": {},
+
+		"broken-filter": {
+			filter: filterBrokenType,
+
+			expectCreateErr: "compile class filter CEL expression: must evaluate to bool",
+		},
+
+		"broken-request": {
+			requests: []*resourceapi.NamedResourcesRequest{requestBrokenType},
+
+			expectCreateErr: "compile request CEL expression: must evaluate to bool",
+		},
+
+		"no-resources": {
+			filter:   filterAny,
+			requests: []*resourceapi.NamedResourcesRequest{requestAny},
+
+			expectAllocateErr: "insufficient resources",
+		},
+
+		"okay": {
+			model:    oneInstance,
+			filter:   filterAny,
+			requests: []*resourceapi.NamedResourcesRequest{requestAny},
+
+			expectAllocation: []string{instance1},
+		},
+
+		"filter-mismatch": {
+			model:    oneInstance,
+			filter:   filterNone,
+			requests: []*resourceapi.NamedResourcesRequest{requestAny},
+
+			expectAllocateErr: "insufficient resources",
+		},
+
+		"request-mismatch": {
+			model:    oneInstance,
+			filter:   filterAny,
+			requests: []*resourceapi.NamedResourcesRequest{requestNone},
+
+			expectAllocateErr: "insufficient resources",
+		},
+
+		"many": {
+			model:    twoInstances,
+			filter:   filterAny,
+			requests: []*resourceapi.NamedResourcesRequest{requestAny, requestAny},
+
+			expectAllocation: []string{instance1, instance2},
+		},
+
+		"too-many": {
+			model:    oneInstance,
+			filter:   filterAny,
+			requests: []*resourceapi.NamedResourcesRequest{requestAny, requestAny},
+
+			expectAllocateErr: "insufficient resources",
+		},
+
+		"filter-evaluation-error": {
+			model:    oneInstance,
+			filter:   filterBrokenEvaluation,
+			requests: []*resourceapi.NamedResourcesRequest{requestAny},
+
+			expectAllocateErr: "evaluate filter CEL expression: no such key: no-such-attribute",
+		},
+
+		"request-evaluation-error": {
+			model:    oneInstance,
+			filter:   filterAny,
+			requests: []*resourceapi.NamedResourcesRequest{requestBrokenEvaluation},
+
+			expectAllocateErr: "evaluate request CEL expression: no such key: no-such-attribute",
+		},
+	}
+
+	for name, tc := range testcases {
+		t.Run(name, func(t *testing.T) {
+			tCtx := ktesting.Init(t)
+
+			controller, allocateErr := NewClaimController(tc.filter, tc.requests)
+			if allocateErr != nil {
+				if tc.expectCreateErr == "" {
+					tCtx.Fatalf("unexpected create error: %v", allocateErr)
+				}
+				require.Equal(tCtx, tc.expectCreateErr, allocateErr.Error())
```

Comment:
Isn't that a weaker check that what I have right now? I want to test the error explanation to ensure that it is readable and came from the right code path, not the error type.

This package doesn't have pre-defined errors, so `errors.Is` has nothing to compare against.

Reply:
> Isn't that a weaker check that what I have right now? 

It depends on the implementation of `Is`.

Reply:
Let's take an example. I want to check that the unknown attribute is included in the error message:
> `evaluate filter CEL expression: no such key: no-such-attribute`

This error is the result of `fmt.Errorf`. Are you suggesting that I should use a custom error type with an `error.Is` implementation which compares the error text? That seems a bit unusual to me. Nothing in the code depends on that behavior, it is only for testing.

Reply:
Or are you suggesting that the unit tests are too strict and shouldn't cover specific error texts? Then the test cases only specify a boolean for "expect some error".

Both approaches are used in various places. Some tests check that an error contains some substring. I'm leaning towards testing a bit more and checking that the full result of formatting and wrapping looks good, but I don't feel strongly about this.

Reply:
> Or are you suggesting that the unit tests are too strict and shouldn't cover specific error texts?

Kind of. We shouldn't need to update the unit tests if we change error messages a little. Is there more than one error possible? If not, maybe the "boolean" approach is ok.

Otherwise, we could define an error type (maybe called `CatchAll`) in some util package that implements `Is` returning true for any error. And once it's necessary to differentiate, the developer can add their own type or pointer to a specific error.

> Both approaches are used in various places. Some tests check that an error contains some substring

Yeah, I've trying to get rid of those, but there are just too many. In the past, comparing errors was harder because `Is` and `%w` didn't exist.

Reply:
Btw, the above intent is documented here https://github.com/kubernetes/community/blob/master/sig-scheduling/CONTRIBUTING.md#technical-and-style-guidelines

Reply:
> Is there more than one error possible?

There is only a single error type. Only the explanation is different.

> If not, maybe the "boolean" approach is ok.

I've gone for that.